File: src/main/java/fr/adrienbrault/idea/symfony2plugin/dic/inspection/ContainerSettingDeprecatedInspection.java
package fr.adrienbrault.idea.symfony2plugin.dic.inspection;

import com.intellij.codeInspection.LocalInspectionTool;
import com.intellij.patterns.ElementPattern;
import com.intellij.codeInspection.ProblemHighlightType;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiElementVisitor;
import com.intellij.psi.xml.XmlAttribute;
import com.intellij.psi.xml.XmlTag;
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent;
import fr.adrienbrault.idea.symfony2plugin.config.yaml.YamlElementPatternHelper;
import fr.adrienbrault.idea.symfony2plugin.util.PsiElementUtils;
import fr.adrienbrault.idea.symfony2plugin.util.psi.PsiElementAssertUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.yaml.psi.YAMLKeyValue;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class ContainerSettingDeprecatedInspection {

    public static class ContainerSettingDeprecatedInspectionYaml extends LocalInspectionTool {
        public @NotNull PsiElementVisitor buildVisitor(final @NotNull ProblemsHolder holder, boolean isOnTheFly) {
            if (!Symfony2ProjectComponent.isEnabled(holder.getProject())) {
                return super.buildVisitor(holder, isOnTheFly);
            }

            return new MyYamlPsiElementVisitor(holder);
        }

        private static class MyYamlPsiElementVisitor extends PsiElementVisitor {
            @NotNull private final ProblemsHolder holder;
            private ElementPattern<PsiElement> insideServiceKeyPattern;

            MyYamlPsiElementVisitor(@NotNull ProblemsHolder holder) {
                this.holder = holder;
            }

            @Override
            public void visitElement(@NotNull PsiElement element) {
                if (element instanceof YAMLKeyValue) {
                    registerYmlRoutePatternProblem(holder, (YAMLKeyValue) element, getInsideServiceKeyPattern());
                }

                super.visitElement(element);
            }

            private ElementPattern<PsiElement> getInsideServiceKeyPattern() {
                return insideServiceKeyPattern != null ? insideServiceKeyPattern : (insideServiceKeyPattern = YamlElementPatternHelper.getInsideServiceKeyPattern());
            }
        }
    }

    public static class ContainerSettingDeprecatedInspectionXml extends LocalInspectionTool {
        public @NotNull PsiElementVisitor buildVisitor(final @NotNull ProblemsHolder holder, boolean isOnTheFly) {
            if (!Symfony2ProjectComponent.isEnabled(holder.getProject())) {
                return super.buildVisitor(holder, isOnTheFly);
            }

            return new PsiElementVisitor() {
                @Override
                public void visitElement(@NotNull PsiElement element) {
                    if (element instanceof XmlAttribute) {
                        registerXmlAttributeProblem(holder, (XmlAttribute) element);
                    }

                    super.visitElement(element);
                }
            };
        }
    }

    private static void registerYmlRoutePatternProblem(@NotNull ProblemsHolder holder, @NotNull YAMLKeyValue element, @NotNull ElementPattern<PsiElement> insideServiceKeyPattern) {
        String s = PsiElementUtils.trimQuote(element.getKeyText());
        if (("factory_class".equals(s) || "factory_method".equals(s) || "factory_service".equals(s)) && insideServiceKeyPattern.accepts(element)) {
            // services:
            //   foo:
            //      factory_*:
            registerProblem(holder, element.getKey());
        }
    }

    private static void registerXmlAttributeProblem(@NotNull ProblemsHolder holder, @NotNull XmlAttribute xmlAttribute) {
        String name = xmlAttribute.getName();
        if (!("factory-class".equals(name) || "factory-method".equals(name) || "factory-service".equals(name))) {
            return;
        }

        XmlTag xmlTagRoute = PsiElementAssertUtil.getParentOfTypeWithNameOrNull(xmlAttribute, XmlTag.class, "service");
        if (xmlTagRoute != null) {
            registerProblem(holder, xmlAttribute.getFirstChild());
        }
    }

    private static void registerProblem(@NotNull ProblemsHolder holder, @Nullable PsiElement target) {
        if (target == null) {
            return;
        }

        holder.registerProblem(target, "Symfony: this factory pattern is deprecated use 'factory' instead", ProblemHighlightType.LIKE_DEPRECATED);
    }
}

File: src/main/kotlin/fr/adrienbrault/idea/symfony2plugin/dic/inspection/ContainerSettingDeprecatedInspection.kt
package fr.adrienbrault.idea.symfony2plugin.dic.inspection

import com.intellij.codeInspection.LocalInspectionTool
import com.intellij.codeInspection.ProblemHighlightType
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.patterns.ElementPattern
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiElementVisitor
import com.intellij.psi.xml.XmlAttribute
import com.intellij.psi.xml.XmlTag
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent
import fr.adrienbrault.idea.symfony2plugin.config.yaml.YamlElementPatternHelper
import fr.adrienbrault.idea.symfony2plugin.util.PsiElementUtils
import fr.adrienbrault.idea.symfony2plugin.util.psi.PsiElementAssertUtil
import org.jetbrains.yaml.psi.YAMLKeyValue

private fun registerYmlRoutePatternProblem(holder: ProblemsHolder, element: YAMLKeyValue, insideServiceKeyPattern: ElementPattern<PsiElement>) {
    val s = PsiElementUtils.trimQuote(element.keyText)
    if ((s == "factory_class" || s == "factory_method" || s == "factory_service") && insideServiceKeyPattern.accepts(element)) {
        // services:
        //   foo:
        //      factory_*:
        registerProblem(holder, element.key)
    }
}

private fun registerXmlAttributeProblem(holder: ProblemsHolder, xmlAttribute: XmlAttribute) {
    val name = xmlAttribute.name
    if (!(name == "factory-class" || name == "factory-method" || name == "factory-service")) {
        return
    }

    val xmlTagRoute = PsiElementAssertUtil.getParentOfTypeWithNameOrNull(xmlAttribute, XmlTag::class.java, "service")
    if (xmlTagRoute != null) {
        registerProblem(holder, xmlAttribute.firstChild)
    }
}

private fun registerProblem(holder: ProblemsHolder, target: PsiElement?) {
    if (target == null) {
        return
    }

    holder.registerProblem(target, "Symfony: this factory pattern is deprecated use 'factory' instead", ProblemHighlightType.LIKE_DEPRECATED)
}

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
open class ContainerSettingDeprecatedInspection {
    open class ContainerSettingDeprecatedInspectionYaml : LocalInspectionTool() {
        override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
            if (!Symfony2ProjectComponent.isEnabled(holder.project)) {
                return super.buildVisitor(holder, isOnTheFly)
            }

            return MyYamlPsiElementVisitor(holder)
        }

        private class MyYamlPsiElementVisitor(private val holder: ProblemsHolder) : PsiElementVisitor() {
            private var insideServiceKeyPattern: ElementPattern<PsiElement>? = null

            override fun visitElement(element: PsiElement) {
                if (element is YAMLKeyValue) {
                    registerYmlRoutePatternProblem(holder, element, getInsideServiceKeyPattern())
                }

                super.visitElement(element)
            }

            private fun getInsideServiceKeyPattern(): ElementPattern<PsiElement> {
                val pattern = insideServiceKeyPattern ?: YamlElementPatternHelper.getInsideServiceKeyPattern()
                insideServiceKeyPattern = pattern
                return pattern
            }
        }
    }

    open class ContainerSettingDeprecatedInspectionXml : LocalInspectionTool() {
        override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
            if (!Symfony2ProjectComponent.isEnabled(holder.project)) {
                return super.buildVisitor(holder, isOnTheFly)
            }

            return object : PsiElementVisitor() {
                override fun visitElement(element: PsiElement) {
                    if (element is XmlAttribute) {
                        registerXmlAttributeProblem(holder, element)
                    }

                    super.visitElement(element)
                }
            }
        }
    }
}
